ggml-qnn.cpp

```diff
@@ -0,0 +1,4833 @@
+    } else {
+        _pfn_rpc_mem_free(_rpcmem_store_map[buf]);
+        _rpcmem_store_map.erase(buf);
+    }
+}
+
+
+int32_t qnn_instance::rpcmem_to_fd(void *buf) {
+    int32_t mem_fd = -1;
+    if (!is_rpcmem_initialized()) {
+        QNN_LOG_WARN("rpc memory not initialized\n");
+    } else {
+        mem_fd = _pfn_rpc_mem_to_fd(buf);
+    }
+
+    return mem_fd;
+}
+
+
+int qnn_instance::register_rpcmem(void * p_data, Qnn_Tensor_t * p_tensor) {
+    if (nullptr == p_data || (nullptr == p_tensor)) {
+        QNN_LOG_WARN("invalid param\n");
+        return 1;
+    }
+
+    if (!is_rpcmem_initialized()) {
+        QNN_LOG_WARN("rpc memory not initialized\n");
+        return 2;
+    }
+
+    if (is_rpcmem_allocated(p_data)) {
+        QNN_LOG_WARN("rpc memory already allocated\n");
+        //return 3;
+    }
+    if (is_rpcmem_registered((QNN_VER_PTR(*p_tensor)->memHandle))) {
+        QNN_LOG_WARN("tensor %s has been registered shared memory\n", (QNN_VER_PTR(*p_tensor)->name));
+        return 4;
+    }
+
+    int32_t mem_fd = rpcmem_to_fd(p_data);
+    if (-1 == mem_fd) {
+        QNN_LOG_WARN("failed to get file descriptor\n");
+        return 5;
+    }
+    QNN_LOG_DEBUG("mem_fd %d\n", mem_fd);
+    Qnn_MemDescriptor_t descriptor = {
+            {QNN_VER_PTR(*p_tensor)->rank, QNN_VER_PTR(*p_tensor)->dimensions, nullptr},
+            QNN_VER_PTR(*p_tensor)->dataType,
+            QNN_MEM_TYPE_ION,
+            {{mem_fd}}};
+    Qnn_MemHandle_t handle = nullptr;
+    int error = QNN_SUCCESS;
+    error = _qnn_interface.qnn_mem_register(
+            _qnn_context_handle,
+            &descriptor,
+            /*numDescriptors=*/1,
+            &handle);
+    if (error != QNN_SUCCESS) {
+        QNN_LOG_WARN("failed to register shared memory, error %d, %s\n", QNN_GET_ERROR_CODE(error),
+              strerror(error));
+        return 6;
+    } else {
+        QNN_LOG_INFO("tensor %s successfully register shared memory\n", (QNN_VER_PTR(*p_tensor)->name));
+    }
+    QNN_VER_PTR(*p_tensor)->memHandle = handle;
+    _qnn_mem_set.insert(handle);
+
+    return 0;
+}
+
+
+void qnn_instance::unregister_rpcmem() {
+    Qnn_ErrorHandle_t error = QNN_SUCCESS;
+
+    if (_qnn_mem_set.empty()) {
+        QNN_LOG_WARN("no rpcmem registered\n");
+    }
+
+    for (auto &mem_handle : _qnn_mem_set) {
+        error = _qnn_interface.qnn_mem_de_register(&mem_handle, 1);
+        if (error != QNN_SUCCESS) {
+            QNN_LOG_WARN("failed to unregister shared memory, error %d\n", QNN_GET_ERROR_CODE(error));
+        }
+    }
+    _qnn_mem_set.clear();
+}
+
+
+bool qnn_instance::is_rpcmem_allocated(void * buf) {
+    return _rpcmem_store_map.count(buf) != 0U;
+}
+
+
+int qnn_instance::load_backend(std::string & lib_path, const QnnSaver_Config_t ** saver_config) {
+    Qnn_ErrorHandle_t error = QNN_SUCCESS;
+    QNN_LOG_DEBUG("lib_path:%s\n", lib_path.c_str());
+
+    void *lib_handle = dlopen(lib_path.c_str(), RTLD_NOW | RTLD_GLOBAL);
```

Comment:
Could you also support Windows/Linux?
You could test with QNN SDK x86_64-linux-clang or x86_64-windows-msvc without real device

P.S. You can refer to my code
https://github.com/mybigday/llama.node/blob/575219e4c51848c8d12abfcd9c691b533052b837/src/ggml-qnn/ggml-qnn.cpp#L2085-L2097

But I'm not sure why my result is very weird (seems the result tensor is empty).
And HTP couldn't inference, when the shared operator got the tensor size differed with the previous

Reply:
@hans00 I'm having the same branch. I cloned this branch and ported it to Windows (i.e. LoadLibrary() etc). I have a ARM64 Windows Laptop with the Snapdragon 8cx and I was wondering if I could fast inference with llama.cpp.

The NN execution goes correctly during the prompt evaluation, but when it's time to start generating new tokens it fails with mismatch dimensions :(

Reply:
Since 2019, not 2024 :)

Reply:
> Since 2019, not 2024 😊

thanks for your correction.